src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -3131,13 +3131,47 @@ trait Types
        */
       def unifyFull(tpe: Type): Boolean = {
         def unifySpecific(tp: Type) = {
-          sameLength(typeArgs, tp.typeArgs) && {
+          if(sameLength(typeArgs, tp.typeArgs)) {
             val lhs = if (isLowerBound) tp.typeArgs else typeArgs
             val rhs = if (isLowerBound) typeArgs else tp.typeArgs
             // This is a higher-kinded type var with same arity as tp.
             // If so (see SI-7517), side effect: adds the type constructor itself as a bound.
             isSubArgs(lhs, rhs, params, AnyDepth) && { addBound(tp.typeConstructor); true }
-          }
+          } else if(settings.YpartialUnification && typeArgs.lengthCompare(0) > 0 && compareLengths(typeArgs, tp.typeArgs) < 0) {
+            // Simple algorithm as suggested by Paul Chiusano in the comments on SI-2712
+            //
+            //   https://issues.scala-lang.org/browse/SI-2712?focusedCommentId=61270
+            //
+            // Treat the type constructor as curried and partially applied, we treat a prefix
+            // as constants and solve for the suffix. For the example in the ticket, unifying
+            // M[A] with Int => Int this unifies as,
+            //
+            //   M[t] = [t][Int => t]
+            //   A = Int
+            //
+            // A more "natural" unifier might be M[t] = [t][t => t]. There's lots of scope for
+            // experimenting with alternatives here.
+
+            val tpSym = tp.typeSymbolDirect
+
+            val numAbstracted = typeArgs.length
+            val numCaptured = tp.typeArgs.length-numAbstracted
+            val (captured, abstracted) = tp.typeArgs.splitAt(numCaptured)
+
+            val lhs = if (isLowerBound) abstracted else typeArgs
+            val rhs = if (isLowerBound) typeArgs else abstracted
+            // This is a higher-kinded type var with same arity as tp.
+            // If so (see SI-7517), side effect: adds the type constructor itself as a bound.
+            isSubArgs(lhs, rhs, params, AnyDepth) && {
+              val absSyms = tpSym.typeParams.drop(numCaptured)
+              val freeSyms = absSyms.map(_.cloneSymbol(tpSym))
```

Comment:
I would combine this line with the above one. No need for `absSyms`, so let's not leave them lying around for someone else to plug them in somewhere they don't belong :-)